packages/gatsby-source-contentful/src/rich-text.js

```diff
@@ -0,0 +1,184 @@
+const _ = require(`lodash`)
+const { BLOCKS, INLINES } = require(`@contentful/rich-text-types`)
+
+const isEntryReferenceNode = node =>
+  [
+    BLOCKS.EMBEDDED_ENTRY,
+    INLINES.ENTRY_HYPERLINK,
+    INLINES.EMBEDDED_ENTRY,
+  ].indexOf(node.nodeType) >= 0
+
+const isAssetReferenceNode = node =>
+  [BLOCKS.EMBEDDED_ASSET, INLINES.ASSET_HYPERLINK].indexOf(node.nodeType) >= 0
+
+const isEntryReferenceField = field => _.get(field, `sys.type`) === `Entry`
+const isAssetReferenceField = field => _.get(field, `sys.type`) === `Asset`
+
+const getFieldProps = (contentType, fieldName) =>
+  contentType.fields.find(({ id }) => id === fieldName)
+
+const getAssetWithFieldLocalesResolved = ({ asset, getField }) => {
+  return {
+    ...asset,
+    fields: _.mapValues(asset.fields, getField),
+  }
+}
+
+const getFieldWithLocaleResolved = ({
+  field,
+  contentTypesById,
+  getField,
+  defaultLocale,
+  resolvedEntryIDs,
+}) => {
+  // If the field is itself a reference to another entry, recursively resolve
+  // that entry's field locales too.
+  if (isEntryReferenceField(field)) {
+    if (resolvedEntryIDs.indexOf(field.sys.id) >= 0) {
+      return field
+    }
+
+    return getEntryWithFieldLocalesResolved({
+      entry: field,
+      contentTypesById,
+      getField,
+      defaultLocale,
+      resolvedEntryIDs: [...resolvedEntryIDs, field.sys.id],
```

Comment:
@pvdz re: your comment [here](https://github.com/gatsbyjs/gatsby/pull/14616#discussion_r354262735): I don't have a Contentful installation with 100k records to test this on... hm. Should we really mutate in place, though? I was trying to go with a purely functional approach, but if the perf costs could be that big, I suppose it'd make sense to bend the rules ;)

Reply:
You could try to clone and get  https://github.com/tsriram/ifsc to work. That's 150k pages.

(Note that it's best if you pin Gatsby to 2.19.17 right now as there was a small regression for this site and .17 is the fastest until that's addressed.)

Reply:
@pvdz but that repo is using CSVs, not Contentful 🤔

Reply:
I can also simply mutate the array if that's safer! (One thing though — there's no performance difference between spreading and using `.concat`, right?)

Reply:
That is true :) The others are not contentful either. I thought, maybe you could convert them to contentful, since they're text based and generate pages. I haven't made a contentful site myself so not sure where the limits are on random content generation (for 100k).

Reply:
As for spread vs concat, I would kind of expect `resolvedEntryIDs.slice(0).concat([field.sys.id])` to be more efficient, but I'm not sure as it's very engine dependent.

As mentioned elsewhere, I would opt for a mutable Set, where you add/remove the entry.